doc/design/persistent_volumes_etcd_data.md

```diff
@@ -0,0 +1,125 @@
+* k8s nodes scheduling a majority of etcd members pods reboots and come back before node controller pod eviction triggers
+
+##### Effect
+
+The pods will go in failed state, the etcd cluster becomes unquorate. Etcd-operator can only restore the cluster from a backup.
+
+#### Case D
+
+##### Event
+
+* k8s nodes executing a majority of etcd members pods becomes partitioned (network partitioning, goes down, etc...) and cannot talk to the api server, after some time the node controller will evict their pods.
+
+##### Effect
+
+The pods will go in an unknown/failed state and eventually be removed from the API server, the etcd cluster becomes unquorate. Etcd-operator can only restore the cluster from a backup.
+
+
+#### Reasons
+
+Currently etcd operator handles cases A and B but when a cluster becomes unquorate (cases C and D) the current option is to restore the cluster from a backup.
+
+User may desire consistency over availability and prefer to wait for the cluster the return in a quorate state instead of restoring from a backup (that will contain old data).
+
+This proposal aims to fix cases C and D permitting to the cluster the return in a quorate state without the need to restore from a backup.
+
+## Proposed Changes
+
+#### Changes
+
+The change to be effective and cover the above failure scenario is split in two parts.
+
+### Part 1
+
+- Add to the cluster spec PodPolicy an option to enable putting etcd data inside a Persistent Volume
+- If podPolicy has persistent volumes enabled, when creating a new member also create a Persistent Volume Claim and use it as the volume source for etcd-data
+- Handle etcd data persistent volume claims cleanup:
+  - set on the PVC an owner like already done with other resources.
+  - in the reconcile loop, remove PVCs that don't have a matching etcd member in the etcd cluster and their related pod doesn't exists in the k8s API
+  - garbage collect also PVCs
+
+#### Pod name to Persistent Volume Claim name mapping
+
+The pod and related PVC are mapped 1 to 1 so their names be easily generated and deducted. I.E. if the pod is called `etcd-0000` its related PVC name will be `etcd-0000-pvc`
+
+#### Results
+
+These changes will put the basis for initial persistent volume management
+
+#### Problems
+
+This first part won't fix case C and D.
+
+### Part 2
+
+#### Changes
+
+- Change reconcile logic to handle deleted pods that had a persistent volume:
+
+If a pod of an etcd cluster member is deleted and it had a persistent volume claim associated, instead of removing the etcd member from the etcd cluster and adding a new member just recreate it (keeping its pod name) binding the existing persistent volume claim. The pod recreation will happen only if the pod doesn't exists anymore in the API because we'll keep the same pod name. This also enable a MANDATORY property: at most once pod existence logic to avoid having in the API at the same time two pods with associated the same PVC.
```

Comment:
We add a member to the etcd-cluster, a new memberId is generated. If we re-use the PVC, the memberId of the failed member and the new member binding to the PV will be differnent. Would this create further problems?

Reply:
it'll be the same member since it has the same data and same startup command, only a new pod since the previous pod has gone (deleted from the api server)

Reply:
Oh. You wouldn't need to add a member as was the case in addOneMember() in cluster/reconcile.go
```go
resp, err := etcdcli.MemberAdd(ctx, []string{newMember.PeerURL()})
```

Reply:
@georgekuruvillak 

> When a member is added to the etcd-cluster, a new memberId is generated.

No new ID for this new change.

> If the PVC is re-used, the memberId of the failed member and the new member binding to the PV will be differnent. 

It would be the same identity, network, state with PV like EBS, GCE PD.